akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ShardingProducerController.scala

```diff
@@ -0,0 +1,429 @@
+      : Option[ActorRef[DurableProducerQueue.Command[A]]] = {
+
+    durableQueueBehavior.map { b =>
+      val ref = context.spawn(b, "durable")
+      context.watch(ref) // FIXME handle terminated, but it's not supposed to be restarted so death pact is alright
+      askLoadState(context, Some(ref), attempt = 1)
+      ref
+    }
+  }
+
+  private def askLoadState[A: ClassTag](
+      context: ActorContext[InternalCommand],
+      durableQueue: Option[ActorRef[DurableProducerQueue.Command[A]]],
+      attempt: Int): Unit = {
+    implicit val loadTimeout: Timeout = 3.seconds // FIXME config
+    durableQueue.foreach { ref =>
+      context.ask[DurableProducerQueue.LoadState[A], DurableProducerQueue.State[A]](
+        ref,
+        askReplyTo => DurableProducerQueue.LoadState[A](askReplyTo)) {
+        case Success(s) => LoadStateReply(s)
+        case Failure(_) => LoadStateFailed(attempt) // timeout
+      }
+    }
+  }
+
+}
+
+class ShardingProducerController[A: ClassTag](
+    context: ActorContext[ShardingProducerController.InternalCommand],
+    producerId: String,
+    producer: ActorRef[ShardingProducerController.RequestNext[A]],
+    msgAdapter: ActorRef[ShardingEnvelope[A]],
+    region: ActorRef[ShardingEnvelope[ConsumerController.SequencedMessage[A]]],
+    durableQueue: Option[ActorRef[DurableProducerQueue.Command[A]]]) {
+  import ShardingProducerController._
+  import DurableProducerQueue.StoreMessageSent
+  import DurableProducerQueue.StoreMessageSentAck
+  import DurableProducerQueue.StoreMessageConfirmed
+  import DurableProducerQueue.MessageSent
+
+  private val requestNextAdapter: ActorRef[ProducerController.RequestNext[A]] =
+    context.messageAdapter(WrappedRequestNext.apply)
+
+  private def active(s: State[A]): Behavior[InternalCommand] = {
+
+    def onMessage(
+        entityId: EntityId,
+        msg: A,
+        replyTo: Option[ActorRef[Done]],
+        totalSeqNr: TotalSeqNr,
+        newReplyAfterStore: Map[TotalSeqNr, ActorRef[Done]]): Behavior[InternalCommand] = {
+
+      val outKey = s"$producerId-$entityId"
+      val newState =
+        s.out.get(outKey) match {
+          case Some(out @ OutState(_, _, Some(nextTo), _, _, _)) =>
+            // there is demand, send immediately
+            send(msg, outKey, out.seqNr, nextTo)
+            val newUnconfirmed = out.unconfirmed :+ Unconfirmed(totalSeqNr, out.seqNr, replyTo)
+            s.copy(
+              out = s.out.updated(outKey, out.copy(seqNr = out.seqNr + 1, nextTo = None, unconfirmed = newUnconfirmed)),
+              replyAfterStore = newReplyAfterStore)
+          case Some(out @ OutState(_, _, None, buffered, _, _)) =>
+            // no demand, buffer
+            // FIXME limit the buffers.
+            context.log.info("Buffering message to entityId [{}], buffer size [{}]", entityId, buffered.size + 1)
+            val newBuffered = buffered :+ Buffered(totalSeqNr, msg, replyTo)
```

Comment:
Hmm, wouldn't it be better with per entity demand protocol of some sort, rather than a generic we can probably send to the shard so give me next and then ending up sticking it in a buffer here?

Reply:
Ah, I missed out on the comment about allowing to send even when the entity isn't started so there is no demand yet.

Reply:
and in the `RequestNext` there is additional demand information so a cooperative producer can use that to play nice, e.g. slow down or reject.
```
      entitiesWithDemand: Set[EntityId],
      bufferedForEntitiesWithoutDemand: Map[EntityId, Int]
```